### PR TITLE
V0.2 vis new

### PR DESCRIPTION
**Pull request by micli**
# Description

This PR merged code from main branch.  It compitable with wenlei's output. This branch added new feature about how to dump runtime data from model. It used for analysis model behaviors and investigate how data changes.

## Linked issue(s)/Pull request(s)

- [issue_number](issue_link)

## Type of Change

- [ ] Non-breaking bug fix
- [ ] Breaking bug fix
- [x] New feature
- [ ] Test
- [ ] Doc update
- [ ] Docker update

## Related Component

- [x] Simulation toolkit
- [ ] RL toolkit
- [ ] Distributed toolkit

## Has Been Tested

- OS:
  - [ ] Windows
  - [x] Mac OS
  - [x] Linux
- Python version:
  - [x] 3.6
  - [x] 3.7
- Key information snapshot(s):
  
## Needs Follow Up Actions

- [x] New release package
- [ ] New docker image

## Checklist

- [x] Add/update the related comments
- [ ] Add/update the related tests
- [ ] Add/update the related documentations
- [ ] Update the dependent downstream modules usage

### CODE REVIEW

File: maro/data_lib/dump_csv_converter.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+    def __init__(self, parent_path="", scenario_name="", prefix="epoch_", serial=0):
+        super().__init__()
+        self._parent_path = parent_path
+        self._prefix = prefix
+        self._serial = serial
+        self._scenario_name = scenario_name
+        self._manifest_created = False
+        self._mapping_created = False
+
+    def __generate_new_folder(self, parent_path):
```
Comment (romic7x) [inline maro/data_lib/dump_csv_converter.py:30]: correct me if I'm wrong, I thought we should avoid using double underscores to specify an object in development

File: maro/simulator/core.py [hunk 2]
```diff
@@ -92,6 +101,17 @@ def reset(self):
 
         self._event_buffer.reset()
 
+        if "enable-dump-snapshot" in self._additional_options:
+            if self._business_engine._frame is not None:
```
Comment (romic7x) [inline maro/simulator/core.py:105]: in one line?

File: maro/data_lib/dump_csv_converter.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+            for file in files:
+                if file.endswith(".meta"):
+                    col_info_dict = self.get_column_info(os.path.join(curDir, file))
+                    data = np.load(os.path.join(curDir, file.replace(".meta", ".npy")))
+
+                    frame_idx = 0
+                    csv_data = []
+                    for frame in data:
+                        node_idx = 0
+                        the_name = file.replace(".meta", "")
```
Comment (kaiqli) [inline maro/data_lib/dump_csv_converter.py:75]: the_name is little strange, may be file_name or else.

File: maro/backends/np_backend.pyx [hunk 4]
```diff
@@ -261,13 +261,22 @@ cdef class NumpyBackend(BackendAbc):
...
         for node_name, data_arr in self._node_data_dict.items():
             # we have to reset by each attribute
             for attr_info in self._node_attr_dict[node_name]:
                 # we only reset frame here, without snapshot list
                 data_arr[0][attr_info.name] = 0
 
+    cdef void dump(self, str filePath):
+        for node_name, data_arr in self._node_data_dict.items():
+            filename = os.path.join(filePath, node_name + ".npy")
+            descFilename = os.path.join(filePath, node_name + ".meta")
```
Comment (romic7x) [inline maro/backends/np_backend.pyx:274]: camalCase?

File: maro/data_lib/dump_csv_converter.py [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+        cols = str.split(columns, ",")
+        element_list = str.split(elements, ",")
+        i = 0
+        for col in cols:
+            col_dict[col] = element_list[i]
+            i = i + 1
+        return col_dict
+
+    def clear_raw_data(self):
+        for curDir, dirs, files in os.walk(self._foldername):
```
Comment (romic7x) [inline maro/data_lib/dump_csv_converter.py:116]: same here

File: maro/data_lib/dump_csv_converter.py [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+                    f.close()
+            else:
+                # copy file to folder.
+                if os.path.exists(filesource):
+                    copyfile(filesource, file_name)
+            content["mappings"] = os.path.basename(filesource)
+
+        dump_details = {}
+        meta_file_list = []
+        for curDir, dirs, files in os.walk(self._last_snapshot_folder):
```
Comment (romic7x) [inline maro/data_lib/dump_csv_converter.py:189]: same here

Comment (romic7x) [inline maro/data_lib/dump_csv_converter.py:189]: if we do not need curDir and dirs in here, what about replacing it with '_'

File: maro/data_lib/dump_csv_converter.py (hunk 5 above)
Comment (romic7x) [reply maro/data_lib/dump_csv_converter.py:116]: same here with - if we do not need curDir and dirs in here, what about replacing it with '_'

**Review by romic7x — DISMISSED** [review]

**Review by romic7x — APPROVED** [review]

**Review by kaiqli — APPROVED** [review]

**Pull request merged**